# Занятие 1. Часть 2

## Функции и вывод на экран

В дальнейшем вы часто будете использовать _функции_ из Python или из специализированных Python-модулей.

_Функция_ -- это какой-то скрытый от вас программный код, который можно применять для решения конкретной задачи.

Например, познакомимся с функцией `print`:
- нужна для того, чтобы выводить на экран текст / числа / что угодно ещё
- формат вызова: `print(<первое, что хотите вывести на экран>, <второе, что хотите вывести на экран>, ...)`

То, что перечисляется внутри скобок и разделяется запятыми, называется *аргументами* или _входами_ функции

<img src="https://i.ibb.co/12sSpRd/Untitled-Frame-1-11.jpg" alt="Drawing" style="width: 800px;"/>

С помощью Python можно решать огромное количество задач. Мы начнем с очень простых и постепенно будем их усложнять, закончив наш курс небольшим проектом. Если вы уже сталкивались с программированием, то вы помните, что обычно самой первой программой становится вывод "Hello, world". Попробуем сделать это в Python.

In [1]:
print('Hello, world!')
print(1)

Hello, world!
1


Обратите внимание, что "Hello, world!" мы написали в кавычках, а единицу - нет. Это связанно с тем,
что в программировании мы имеем дело с разными типами данных. И Python будет воспринимать текст как текст (строковую переменную), только в том случае, если мы его возьмем в кавычки (неважно, одинарные или двойные). А при выводе эти кавычки отображаться уже не будут (они служат знаком для Python, что внутри них - текст).

print() - это функция, которая выводит то, что мы ей передадим. В других IDE это был бы вывод в терминал, а в Jupyter вывод напечатается под запускаемой ячейкой. Распознать функцию в питоне можно по скобкам после слова, внутри которых мы передаем аргумент, к которому эту функцию нужно применить (текст "Hello, world" или 1 в нашем случае). 

In [2]:
print(Hello, world)

NameError: name 'Hello' is not defined

Написали без кавычек - получили ошибку. Кстати, обратите внимание, что очень часто в тексте ошибки есть указание на то, что именно произошло, и можно попробовать догадаться, что же нужно исправить. Текст без кавычек Python воспринимает как название переменной, которую еще не задали. Кстати, если забыть закрыть или открыть кавычку (или поставить разные), то тоже поймаем ошибку.


Иногда мы хотим комментировать наш код, чтобы я-будущий или наши коллеги поменьше задавались вопросами, а что вообще имелось ввиду. Комментарии можно писать прямо в коде, они не повлияют на работу программы, если их правильно оформить.

In [5]:
# Обратите внимание: так выглядит комментарий - часть скрипта, которая не будет исполнена 
# при запуске программы.
# Каждую строку комментария мы начинаем со знака хэштега.

'''
Это тоже комментарий - обычно выделение тремя апострофами мы используем для тех случаев, 
когда хотим написать длинный, развернутый текст.
'''

print('Hello, world')

Hello, world


Обратите внимание, что в отличие от других IDE (например, PyCharm) в Jupyter Notebook не всегда обязательно использовать print(), чтобы что-то вывести. Но не относитесь к этому как к стандарту для любых IDE.

In [1]:
"Hello, world"

'Hello, world'

Выше рядом с выводом появилась подпись Out[]. В данном случае Jupyter показывает нам последнее значение, лежащее в буфере ячейки. Например, в PyCharm такой вывод всегда будет скрыт, пока мы его не "напечатаем" с помощью print(). Но эта особенность Jupyter помогает нам быстро проверить, что, например, находится внутри переменной, когда мы отлаживаем наш код.

Следующая вещь, которую нужно знать про язык программирования - как в нем задаются переменные. Переменные - это 
контейнеры, которые хранят в себе информацию (текстовую, числовую, какие-то более сложные виды данных). В Python
знаком присвоения является знак =. 

In [1]:
x = 'Hello, world!'
print(x)    # Обратите внимание, что результат вызова этой функции такой же, как выше, 
            # только текст теперь хранится внутри переменной.

Hello, world!


Python -  язык чувствительный к регистру. Поэтому, когда создаете/вызываете переменные или функции, обязательно используйте правильный регистр. Так, следующая строка выдаст ошибку.

In [2]:
print(X) # мы создали переменную x, а X не существует

NameError: name 'X' is not defined

Еще раз обратим внимание на ошибку. *NameError: name 'X' is not defined* означает, что переменная с таким названием не была создана в этой программе. Кстати, обратите внимание, что переменные в Jupyter хранятся на протяжении всей сессии (пока вы работаете с блокнотом и не закрыли его), и могут быть созданы в одной ячейке, а вызваны в другой. Давайте опять попробуем обратиться к x.

In [11]:
print(x) # работает!

Hello, world!


## Типы данных: Целые числа - INT (INTEGER)

Знакомство с типа данных мы начнем с целых чисел. Если вы вдруг знакомы с другими языками программирования, то стоит отметить, что типизация в Python - динамическая. Это значит, что вам не нужно говорить какой тип данных вы хотите положить в переменную - Python сам все определит. Проверить тип данных можно с помощью функции type(), передав ей в качестве аргумента сами данные или переменную.

**ЦЕЛЫЕ ЧИСЛА (INT, INTEGER):** 1, 2, 592, 1030523235 - любое целое число без дробной части. 

In [3]:
y = 2
print(type(2))
print(type(y))

<class 'int'>
<class 'int'>


Обратите внимание - выше мы вызвали функцию внутри функции. 
type(2) возвращает скрытое значение типа переменной (int для integer). 
Чтобы вывести это скрытое значение - мы должны его "напечатать".

Самое элементарное, что можно делать в Python - использовать его как калькулятор. Давайте посмотрим, как
он вычитает, складывает и умножает.

In [14]:
print(2 + 2)
print(18 - 9)
print(4 * 3)

4
9
12


С делением нужно быть немного осторожней. Существует два типа деления - привычное нам, которое даст в ответе дробь при делении 5 на 2, и целочисленное деление, в результате которого мы получим только целую часть частного.

In [4]:
print(5 / 2)  # в результате такого деления получается другой тип данных (float), подробнее о нем поговорим позже.
print(5 // 2)

2.5
2


А если нам надо как раз найти остаток от деления, то мы можем воспользоваться оператором модуло %


In [17]:
print(5 % 2)

1


Еще одна математическая операция, которую мы можем выполнять без загрузки специальных математических библиотек - это
возведение в степень.

In [5]:
print(5**2)

25


Все то же самое работает и с переменными, содержащими числа.

In [19]:
a = 2
b = 3
print(a ** b)
# изменится ли результат, если мы перезапишем переменную a?
a = 5
print(a ** b)

8
125


Часто возникает ситуация, что мы считали какие-то данные в формате текста, и у нас не работают арифметические операции. Тогда мы можем с помощью функции int() преобразовать строковую переменную (о них ниже) в число, если эта строка может быть переведена в число в десятичной системе.

In [6]:
print(2 + '5') # ошибка, не сможем сложить целое число и строку

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [7]:
print(2 + int('5')) # преобразовали строку в число и все заработало

7


In [8]:
int('текст') # здесь тоже поймаем ошибку, потому что строка не представляет собой число

ValueError: invalid literal for int() with base 10: 'текст'

## (∩｀-´)⊃━☆ﾟ.*･｡ﾟ Задача

### Сумма цифр трехзначного числа

Дано трехзначное число 179. Найдите сумму его цифр.

**Формат вывода**  
Выведите ответ на задачу.

**Ответ**  
Вывод программы:  
17

In [9]:
# (∩｀-´)⊃━☆ﾟ.*･｡ﾟ

x = 179
x_1 = x // 100
x_2 = x // 10 % 10
x_3 = x % 10
print(x_1, x_2, x_3) # тестовый вывод, проверяем, что правильно "достали" цифры из числа
print(x_1 + x_2 + x_3) # ответ на задачу

1 7 9
17


## (∩｀-´)⊃━☆ﾟ.*･｡ﾟ Задача

### Электронные часы

Дано число N. С начала суток прошло N минут. Определите, сколько часов и минут будут показывать электронные часы в этот момент.

**Формат ввода**  

Вводится число N — целое, положительное, не превышает 10⁷.

**Формат вывода**

Программа должна вывести два числа: количество часов (от 0 до 23) и количество минут (от 0 до 59).

Учтите, что число N может быть больше, чем количество минут в сутках.

#### Примеры

Тест 1  
**Входные данные:**  
150

**Вывод программы:**  
2 30

In [2]:
# (∩｀-´)⊃━☆ﾟ.*･｡ﾟ

minutes = 150

print(minutes // 60 % 24, minutes % 60)

2 30


## Типы данных: Вещественные числа - FLOAT

На что обратить внимание:
- числа с десятичной частью, например, $1.43$, в отличие от Excel, в Python пишутся не через запятую `,`, а через точку `.`.

Почувствуйте разницу:

In [1]:
print(2,6 / 2)  # печатает два числа
print(2,   6 / 2)  # и так тоже два числа, пробел в данном случае неважен
print(2.6 / 2)  # печатает одно число

2 3.0
2 3.0
1.3


Также не стоит путать целочисленное и вещественное деление:

In [2]:
print(9 / 2)  # печатает ожидаемый дробный результат
print(9 // 2)  # целочисленное деление
print(9 % 2)  # остаток от (целочисленного) отделения

4.5
4
1


In [3]:
print(0.1+0.2)

0.30000000000000004


In [4]:
0.2+0.3

0.5

Еще результат может выводиться в довольно странном формате:

In [5]:
1.543**1000

2.3223413936165768e+188

Такая запись расшифровывается следующим образом:

$ n \cdot e+188 = n \cdot 10^{188} $

## Типы данных: Строки (текст) - STR (STRING)

С помощью `print` можно печатать не только числа. Например, текст:
- текстовая строка в питоне -- это любое выражение, заключённое в двойные `"` или одинарные `'` кавычки
- с точки зрения пользователя ничем не отличается от привычного текста

Например:

In [6]:
print('Моё имя:')

Моё имя:


In [1]:
print('Первое'+' '+'занятие')

Первое занятие


In [8]:
print('У меня '+ 'огромная проблема '*20)

У меня огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема огромная проблема 


На что обратить внимание:
- выражение `1` это число
- выражение `'1'` это строка

Почувствуйте разницу:

In [9]:
print(1 + 2)  # печатает число
print('1' + '2')  # печатает результат склеивания строк
print(str(1) + str(2))  # функция str преобразует аргумент в строку

3
12
12


# Типы данных: логические или булевы переменные (boolean)

Следующий тип данных - это логические переменные. Логические переменные могут принимать всего два значения - **истина (True)** или **ложь (False)**. В Python тип обознчается **bool**.

In [12]:
print(type(True), type(False))

<class 'bool'> <class 'bool'>


Логические переменные чаще всего используется в условных операторах if-else и в цикле с остановкой по условию while. В части по анализу данных еще увидим одно частое применение - создание булевых масок для фильтрации данных (например, вывести только те данные, где возраст больше 20).

Обратите внимание, что True и False обязательно пишутся с большой буквы и без кавычек, иначе можно получить ошибку.

In [22]:
print(type('True')) # тип str - строковая переменная
print(true) # ошибка, Python думает, что это название переменной

<class 'str'>


NameError: name 'true' is not defined

Как и в случае чисел и строк, с логическими переменными работает преобразование типов. Превратить что-либо в логическую переменную можно с помощью функции bool().

Преобразование чисел работает следующим образом - 0 превращается в False, все остальное в True.

In [33]:
print(bool(0))
print(bool(23))
print(bool(-10))

False
True
True


Пустая строка преобразуется в False, все остальные строки в True.

In [34]:
print(bool(''))
print(bool('Hello'))
print(bool(' ')) # даже строка из одного пробела - это True
print(bool('False')) # и даже строка 'False' - это True

False
True
True
True


И при необходимости булеву переменную можно конвертировать в int. Тут все без сюрпризов - ноль и единица.

In [23]:
print(int(True))
print(int(False))

1
0


## Логические выражения

Давайте посмотрим, где используется новый тип данных.

Мы поработаем с логическими выражениями, результат которых - булевы переменные.

Логические выражения выполняют проверку на истинность, то есть выражение равно True, если оно истинно, и False, если ложно.

В логических выражениях используются операторы сравнения:
* ==  (равно)
* !=  (не равно)
* \>  (больше)
* <  (меньше)
* \>=  (больше или равно)
* <=  (меньше или равно)

In [24]:
print(1 == 1)
print(1 != '1') 
c = 1 > 3
print(c)
print(type(c))
x = 5
print(1 < x <= 5) # можно писать связки цепочкой

True
True
False
<class 'bool'>
True


Логические выражения можно комбинировать с помощью следующих логических операторов:

* логическое И (and) - выражение истинно, только когда обе части истинны, иначе оно ложно
* логическое ИЛИ (or) - выражение ложно, только когда обе части ложны, иначе оно истинно
* логическое отрицание (not) - превращает True в False и наоборот

In [37]:
print((1 == 1) and ('1' == 1))
print((1 == 1) or ('1' == 1))
print(not(1 == 1))
print(((1 == 1) or ('1' == 1)) and (2 == 2))

False
True
False
True


## (∩｀-´)⊃━☆ﾟ.*･｡ﾟ Задача
## Вася в Италии

Вася уехал учиться по обмену на один семестр в Италию. Единственный магазин в городе открыт с 6 до 8 утра и с 16 до 17 вечера (включительно). Вася не мог попасть в магазин уже несколько дней и страдает от голода. Он может прийти в магазин в X часов. Если магазин открыт в X часов, то выведите True, а если закрыт - выведите False.

В единственной строке входных данных вводится целое число X, число находится в пределах от 0 до 23

**Формат ввода**  

Целое число X, число находится в пределах от 0 до 23

**Формат вывода**

True или False

#### Примеры

Тест 1  
**Входные данные:**  
16

**Вывод программы:**  
True

In [39]:
## (∩｀-´)⊃━☆ﾟ.*･｡ﾟ 

time = 16
can_visit = 6 <= time <= 8
can_visit2 = 16 <= time <= 17
print(can_visit or can_visit2)

True


## Переменные

В Python мы можем записывать в переменные какие-либо значения (или наборы значений):

In [3]:
MyName
my_name
My_name

NameError: name 'MyName' is not defined

In [4]:
number = 99
my_name = 'Маргарита'

In [5]:
print(number)
print(my_name)

99
Маргарита


Если мы будем реализовывать какие-то действия с переменными, то они будут осуществляться со значениями, которые записаны в эти переменные:

In [6]:
print(number*my_name)

МаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргаритаМаргарита


# Команда input()

Ранее наши программы выводили на экран только фразы, известные в момент написания кода. Но программы могут работать и с данными, которые станут известны только во время выполнения: например, их будет вводить пользователь с клавиатуры. Реализовать это можно так:

In [ ]:
print('Как тебя зовут?')
name = input()
print('Привет,', name)

Команда input() всегда записыввется c круглыми скобками. Работает так: когда программа доходит до места, где есть input(), она ждет, пока пользователь введет строку с клавиатуры (ввод завершается нажатием клавиши Enter). Введенная строка подставляется на место input(). Т.е. input() получает от пользователя какие-то данные и в место вызова подставляет строковое значение, в нашем случае записывает его в качестве значения переменной <b>name</b>. 